core/storage/trie/impl/trie_storage_impl.hpp

```diff
@@ -0,0 +1,64 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_STORAGE_TRIE_IMPL_TRIE_STORAGE_IMPL
+#define KAGOME_STORAGE_TRIE_IMPL_TRIE_STORAGE_IMPL
+
+#include "storage/changes_trie/changes_tracker.hpp"
+#include "storage/trie/codec.hpp"
+#include "storage/trie/impl/polkadot_trie_factory.hpp"
+#include "storage/trie/impl/trie_serializer.hpp"
+#include "storage/trie/trie_storage.hpp"
+
+namespace kagome::storage::trie {
+
+  class TrieStorageImpl : public TrieStorage {
+   public:
+    static outcome::result<std::unique_ptr<TrieStorageImpl>> createEmpty(
+        std::shared_ptr<PolkadotTrieFactory> trie_factory,
+        std::shared_ptr<Codec> codec,
+        std::shared_ptr<TrieSerializer> serializer,
+        boost::optional<std::shared_ptr<changes_trie::ChangesTracker>> changes);
+
+    static outcome::result<std::unique_ptr<TrieStorageImpl>> createFromStorage(
+        const common::Buffer &root_hash,
+        std::shared_ptr<Codec> codec,
+        std::shared_ptr<TrieSerializer> serializer,
+        boost::optional<std::shared_ptr<changes_trie::ChangesTracker>> changes);
+
+    TrieStorageImpl(TrieStorageImpl const &) = delete;
+    void operator=(const TrieStorageImpl &) = delete;
+
+    ~TrieStorageImpl() override = default;
+
+    outcome::result<std::unique_ptr<PersistentTrieBatch>> getPersistentBatch()
+        override;
+    outcome::result<std::unique_ptr<EphemeralTrieBatch>> getEphemeralBatch()
+        const override;
+
+    outcome::result<std::unique_ptr<PersistentTrieBatch>>
+    getPersistentBatchAt(const common::Hash256 &root) override;
+    outcome::result<std::unique_ptr<EphemeralTrieBatch>>
+    getEphemeralBatchAt(const common::Hash256 &root) const override;
+
+    common::Buffer getRootHash() const override;
+
+   protected:
+    TrieStorageImpl(
+        common::Buffer root_hash,
+        std::shared_ptr<Codec> codec,
+        std::shared_ptr<TrieSerializer> serializer,
+        boost::optional<std::shared_ptr<changes_trie::ChangesTracker>> changes);
```

Comment:
Why it is optional? What is the case when we don't need to use changes tracker?

Reply:
Yes. In tests, for example. And I think in Riot I was once told that it's possible that a node doesn't have a changes trie, so might be useful here either.